### PR TITLE
Recipe App

### PR DESCRIPTION
**Pull request by hamidazim321**
# PR Summary
- We have added the following models and associations
    - `food`
    - `recipe`
    - `user`
    - `food_recipe`
    - `inventory_recipe`
    - `inventory`
- We have added user authentication using `devise` and authorization using `CanCanCan` library
- We have added unit tests and integration tests
- We have added all views according to the wireframe
- We have added controller actions for all required methods `PUT`, `PATCH`, `UPDATE`, `DESTROY`, `GET`
- We have added a shopping list view and a modal for shopping list generator
- We have added all the required associations and the `recipe_food` and `inventory_food` joining tables for the many_to_many relations
- We have added a Navigation bar

## Changes after previous review
- I have updated public recipe query and view as per suggested by the reviewer to match the wireframe
- I have added authorization check using `CanCanCan` library before allowing user to delete a recipe item
- I have changed to the `Add food` button on `recipe#show` to `add ingredient` as per suggested by the reviewer
- I have added authorization check using `CanCanCan` library before allowing user to delete a inventory item.
**In summary**: I have applied all the required changes by the previous reviewer

### CODE REVIEW

File: app/views/recipes/public_recipes.html.erb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+              <%= button_to "Remove", recipe_path(recipe.id), method: :delete, data: { confirm: "Are you sure?", turbo_confirm: "Are you sure?" }, class: 'btn btn-danger'%>
+              <%else%>
+              <p>By: <%= recipe.user.name%></p>
+            <%end%>
+        </div>
+        <div class= "w-75">
+          <p><%= recipe.description%></p>
+        </div>
+      </li>
+    <%end%>
```
Comment (zeff96) [inline app/views/recipes/public_recipes.html.erb]: - [ ] Hey team, fantastic job so far—really impressive! 🚀 However, I've noticed a couple of minor issues that need addressing. According to the project requirements, the public shopping list should exclusively feature recipes from other users, showcasing their names and total prices. In the current implementation, there's a missing element: the total prices for the listed recipes are not being displayed. Let's tweak this to align perfectly with the project specifications! 🛠️👩‍🍳

Project design
image

your design
image

Kindly adjust to match the given wireframe

File: app/views/recipes/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<section>
+  <%= link_to "add new recipe", new_recipe_path, class: 'btn btn-success mb-3'%>
+  <ul class="list-unstyled d-flex flex-column gap-3"> 
+    <%@recipes.each do |recipe|%>
+      <li class="d-flex gap-5 border border-2 p-2">
+        <div class="w-25">
+            <%= link_to recipe_path(recipe) do%>
+              <h3 class="fs-3"><%= recipe.name%></h3>
+            <%end%>
+            <%= button_to "Remove", recipe_path(recipe.id), method: :delete, data: { confirm: "Are you sure?", turbo_confirm: "Are you sure?" }, class: 'btn btn-danger'%>
```
Comment (zeff96) [inline app/views/recipes/index.html.erb]: - [ ] 🔐 According to the project guidelines, only the rightful owner should have the authority to delete a recipe item. It's crucial to enhance security by verifying that the user attempting the deletion is indeed the rightful owner. To achieve this, consider implementing an authorization mechanism, such as CancanCan or any other suitable solution. This mechanism will add an additional layer of protection by ensuring that only users with the appropriate permissions can access the delete functionality. This way, we strengthen the security of our application and guarantee that only authorized individuals can modify or remove recipe items. 🚀

File: app/views/recipes/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    <p><strong>Cooking time:</strong> <%= @recipe.cooking_time%> hours</p>
+  </div>
+  <div>
+    <h3>instructions</h3>
+    <p><%= @recipe.description%></p>
+  </div>
+  <div class="d-flex justify-content-between border-bottom border-2 pb-2">
+    <button data-controller="modal" data-modal-id="shopping-list-generator" data-action="click->modal#controlModal">Generate shopping list</button>
+    <%if can? :update, @recipe%>
+      <button class="ms-auto" id="food-btn" data-controller="modal" data-modal-id="food-form" data-action="click->modal#controlModal">add food</button>
```
Comment (zeff96) [inline app/views/recipes/show.html.erb]: - [ ] Hello, as per the given wireframe this button should "Add ingredient"

File: app/views/inventories/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    <% if @inventories.empty? %>
+      <p>No inventories yet.</p>
+    <% else %>
+      <% @inventories.each do |inventory| %>
+        <div class="inventory-card">
+          <div class="inventory-card-header gap-5">
+              <%= link_to inventory.name, inventory_path(inventory.id), class: "inventory-card-name" %>
+              <p class="inventory-card-description text-start"><%= inventory.description %></p> 
+          </div>
+          <%= button_to "Remove", inventory_path(inventory.id), method: :delete, data: { confirm: "Are you sure?", turbo_confirm: "Are you sure?" }, class: "inventory-card-link inventory-card-link-remove text-danger" %>
```
Comment (zeff96) [inline app/views/inventories/index.html.erb]: - [ ] 🔐 According to the project guidelines, only the rightful owner should have the authority to delete a inventory item. It's crucial to enhance security by verifying that the user attempting the deletion is indeed the rightful owner. To achieve this, consider implementing an authorization mechanism, such as CancanCan or any other suitable solution. This mechanism will add an additional layer of protection by ensuring that only users with the appropriate permissions can access the delete functionality. This way, we strengthen the security of our application and guarantee that only authorized individuals can modify or remove inventory items. 🚀

**Review by zeff96 — CHANGES_REQUESTED** [review]
### Status: CHANGES REQUIRED ♻️
Hi @hamidazim321, @hydaline-code and @BilalLiaquat7 , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

almost

## Highlights
:heavy_check_mark: Linters passing
:heavy_check_mark: Correct GitFlow
:heavy_check_mark: All test cases passing

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by thecodechaser — APPROVED** [review]
## STATUS: APPROVED 🟢

## Hi @hamidazim321, @hydaline-code, and @BilalLiaquat7👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight 🎉
- Good job with model validations. ✔️
- Nice UI design. ✔️
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- No linter errors. :heavy_check_mark:
- PR has a good title and summary. ✔️



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
